src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,478 @@
+    /// </remarks>
+    internal abstract class DeterministicKeyBuilder
+    {
+        protected DeterministicKeyBuilder()
+        {
+        }
+
+        protected void WriteFileName(JsonWriter writer, string name, string? filePath, DeterministicKeyOptions options)
+        {
+            if ((options & DeterministicKeyOptions.IgnorePaths) != 0)
+            {
+                filePath = Path.GetFileName(filePath);
+            }
+
+            writer.Write(name, filePath);
+        }
+
+        protected void WriteByteArrayValue(JsonWriter writer, string name, ImmutableArray<byte> value)
+        {
+            if (!value.IsDefault)
+            {
+                WriteByteArrayValue(writer, name, value.AsSpan());
+            }
+        }
+
+        internal static void EncodeByteArrayValue(ReadOnlySpan<byte> value, StringBuilder builder)
+        {
+            foreach (var b in value)
+            {
+                builder.Append(b.ToString("x"));
+            }
+        }
+
+        protected void WriteByteArrayValue(JsonWriter writer, string name, ReadOnlySpan<byte> value)
+        {
+            var builder = PooledStringBuilder.GetInstance();
+            EncodeByteArrayValue(value, builder.Builder);
+            writer.Write(name, builder.ToStringAndFree());
+        }
+
+        protected void WriteType(JsonWriter writer, string key, Type? type)
+        {
+            writer.WriteKey(key);
+            WriteType(writer, type);
+        }
+
+        protected void WriteType(JsonWriter writer, Type? type)
+        {
+            if (type is null)
+            {
+                writer.WriteNull();
+                return;
+            }
+
+            writer.WriteObjectStart();
+            writer.Write("fullName", type.FullName);
+            // Note that the file path to the assembly is deliberately not included here. The file path
+            // of the assembly does not contribute to the output of the program.
+            writer.Write("assemblyName", type.Assembly.FullName);
+            writer.Write("mvid", type.Assembly.ManifestModule.ModuleVersionId.ToString());
+            writer.WriteObjectEnd();
+        }
+
+        private (JsonWriter, PooledStringBuilder) CreateWriter()
+        {
+            var builder = PooledStringBuilder.GetInstance();
+            var writer = new StringWriter(builder);
+            return (new JsonWriter(writer), builder);
+        }
+
+        internal string GetKey(
+            CompilationOptions compilationOptions,
+            ImmutableArray<SyntaxTree> syntaxTrees,
+            ImmutableArray<MetadataReference> references,
+            ImmutableArray<AdditionalText> additionalTexts = default,
+            ImmutableArray<DiagnosticAnalyzer> analyzers = default,
+            ImmutableArray<ISourceGenerator> generators = default,
+            EmitOptions? emitOptions = null,
+            DeterministicKeyOptions options = default)
```

Comment:
we should make thsi cancellable.  it does expensive work (that does take cancellation tokens).  I do this in my branch.